lib/encodings/base64.fz

```diff
@@ -0,0 +1,155 @@
+# Base64 encoding and decoding as defined in RFC 4648
+public base64 is
+
+  # allows redefinition for base64url
+  module get_alphabet => "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".as_codepoints.as_array
+  module get_char_62 u8 => 43 # +
+  module get_char_63 u8 => 47 # /
+  module get_encoding_name => "base64"
+
+  # NYI: compiler does currently not optimize features without arguments to treat them like fields
+  char_62_ascii := get_char_62
+  char_63_ascii := get_char_63
+  alphabet := get_alphabet
+  encoding_name := get_encoding_name
+
+  # Encodes a given byte sequence in base64, output length is padded to multiple of 4
+  public encode (data Sequence u8) String =>
+
+    # 0 -> 0; 1 -> 2; 2 -> 1
+    pad_size := (3 - data.count%3)%3     # length of the padding: number of padded zero bytes in input as well as number of '=' in output
+    in_len := data.count + pad_size      # length of zero padded input data
+
+    # this guarantees that input length is always multiple of 3
+    input := array u8 in_len i->if i < data.count then data[i] else 0
+
+    for b0 := 0, b0 + 3
+        b1 := 1, b1 + 3
+        b2 := 2, b2 + 3
+        res := "", res + block # the encoded input data
+    while b2 < in_len
+      # get 3 bytes (24 bits) and convert them into a number with 24 bits
+      n u32 := (input[b0].as_u32 << 16) + (input[b1].as_u32 << 8) + input[b2].as_u32
+
+      # extract four sextets from 24 bit number
+      n0 => ((n >> 18) & 63).as_i32
+      n1 => ((n >> 12) & 63).as_i32
+      n2 => ((n >> 6) & 63).as_i32
+      n3 => (n & 63).as_i32
+
+      # encode sextets
+      block := alphabet[n0] + alphabet[n1] + alphabet[n2] + alphabet[n3] # NYI: OPTIMIZATION: lots of string concatenation
+    else
+      res.substring 0 (res.byte_length - pad_size) + "=" * pad_size
```

Comment:
I prefer either ("=" * pad_size) or "="*pad_size. This only works because of operator precedence where * is _stronger_ than +, but operator precedence does not really make sense for strings...

Reply:
I would argue that it makes sense in some way. When using the same operators as for numbers it seems intuitive to me that they have the same precedence. Also I've tried to avoid parentheses where possible :)
But I do get your point.

Reply:
Another option (probably slightly better performance) is to already return padding as a string. 
"=", "==", "===" and use padding.byte_length and just `+ padding`